code/game/objects/structures/crates_lockers/closets.dm

```diff
@@ -90,26 +97,65 @@
 	. = new_overlays
-	if(enable_door_overlay)
-		if(opened && has_opened_overlay)
-			. += "[icon_door_override ? icon_door : icon_state]_open"
-			var/mutable_appearance/door_blocker = mutable_appearance(icon, "[icon_door || icon_state]_open", plane = EMISSIVE_PLANE)
-			door_blocker.color = GLOB.em_block_color
-			. += door_blocker // If we don't do this the door doesn't block emissives and it looks weird.
-		else if(has_closed_overlay)
-			. += "[icon_door || icon_state]_door"
+	if(!is_animating_door)
+		if(enable_door_overlay)
+			if(opened && has_opened_overlay)
+				. += "[icon_door_override ? icon_door : icon_state]_open"
+				var/mutable_appearance/door_blocker = mutable_appearance(icon, "[icon_door || icon_state]_open", plane = EMISSIVE_PLANE)
+				door_blocker.color = GLOB.em_block_color
+				. += door_blocker // If we don't do this the door doesn't block emissives and it looks weird.
+			else if(has_closed_overlay)
+				. += "[icon_door || icon_state]_door"
+
+		if(opened)
+			return
 
-	if(opened)
-		return
+		if(welded)
+			. += icon_welded
 
-	if(welded)
-		. += icon_welded
+		if(broken || !secure)
+			return
+		//Overlay is similar enough for both that we can use the same mask for both
+		. += emissive_appearance(icon, "locked", alpha = src.alpha)
+		. += locked ? "locked" : "unlocked"
 
-	if(broken || !secure)
+/obj/structure/closet/proc/animate_door(closing)
+	if(!door_anim_time)
 		return
-	//Overlay is similar enough for both that we can use the same mask for both
-	. += emissive_appearance(icon, "locked", alpha = src.alpha)
-	. += locked ? "locked" : "unlocked"
+	if(!door_obj) door_obj = new
+	vis_contents |= door_obj
+	door_obj.icon = icon
+	door_obj.icon_state = "[icon_door || icon_state]_door"
+	is_animating_door = TRUE
+	var/num_steps = door_anim_time / world.tick_lag
+	for(var/I in 0 to num_steps)
+		var/angle = door_anim_angle * (closing ? 1 - (I/num_steps) : (I/num_steps))
+		var/matrix/M = get_door_transform(angle)
+		var/door_state = angle >= 90 ? "[icon_door_override ? icon_door : icon_state]_back" : "[icon_door || icon_state]_door"
+		var/door_layer = angle >= 90 ? FLOAT_LAYER : ABOVE_MOB_LAYER
+
+		if(I == 0)
+			door_obj.transform = M
+			door_obj.icon_state = door_state
+			door_obj.layer = door_layer
+		else if(I == 1)
+			animate(door_obj, transform = M, icon_state = door_state, layer = door_layer, time = world.tick_lag, flags = ANIMATION_END_NOW)
+		else
+			animate(transform = M, icon_state = door_state, layer = door_layer, time = world.tick_lag)
+	addtimer(CALLBACK(src,.proc/end_door_animation),door_anim_time,TIMER_UNIQUE|TIMER_OVERRIDE)
+
+/obj/structure/closet/proc/end_door_animation()
+	is_animating_door = FALSE
+	vis_contents -= door_obj
+	update_icon()
+	COMPILE_OVERLAYS(src)
+
+/obj/structure/closet/proc/get_door_transform(angle)
+	var/matrix/M = matrix()
+	M.Translate(-door_hinge_x, 0)
+	M.Multiply(matrix(cos(angle), 0, 0, -sin(angle) * door_anim_squish, 1, 0))
+	M.Translate(door_hinge_x, 0)
+	return M
```

Comment:
...

Reply:
Insightful. Do I also have to mention that this is a port?